File: Code/brute_hull.cpp

// To run code open Cmd in folder, type g++ brute_hull.cpp, run executable.

//#include <SFML/Graphics.hpp>
#include <vector>
#include <chrono>
#include <fstream>
#include <sstream>
#include <string>
#include <iostream>

struct Point {
    double x, y;
};

// ----------------------
// Load points from CSV
// ----------------------
std::vector<Point> load_points(const std::string& filename) {
    std::vector<Point> points;
    std::ifstream file(filename);
    std::string line;
    while (std::getline(file, line)) {
        if (line.empty()) continue;
        std::stringstream ss(line);
        double x, y;
        char comma;
        ss >> x >> comma >> y;
        points.push_back({x, y});
    }
    return points;
}

// ----------------------
// Brute-force convex hull
// ----------------------
double cross(const Point& p1, const Point& p2, const Point& p3) {
    return (p2.x - p1.x)*(p3.y - p1.y) - (p2.y - p1.y)*(p3.x - p1.x);
}

std::vector<std::pair<Point, Point>> brute_force_convex_hull(const std::vector<Point>& points) {
    std::vector<std::pair<Point, Point>> edges;
    int n = points.size();
    for (int i = 0; i < n; ++i) {
        for (int j = i+1; j < n; ++j) {
            bool left = false, right = false;
            for (int k = 0; k < n; ++k) {
                if (k == i || k == j) continue;
                double cp = cross(points[i], points[j], points[k]);
                if (cp > 0) left = true;
                if (cp < 0) right = true;
                if (left && right) break;
            }
            if (!(left && right)) edges.push_back({points[i], points[j]});
        }
    }
    return edges;
}

/*/ ----------------------
// Optional visualization
// ----------------------
void visualize(const std::vector<Point>& points, const std::vector<std::pair<Point, Point>>& edges, const std::string& filename) {
    sf::RenderTexture texture;
    const int size = 600;
    texture.create(size, size);
    texture.clear(sf::Color::White);

    // Find max X/Y for scaling
    double maxX = 0, maxY = 0;
    for (auto& p : points) {
        if (p.x > maxX) maxX = p.x;
        if (p.y > maxY) maxY = p.y;
    }
    double scaleX = size / maxX;
    double scaleY = size / maxY;

    // Draw points
    for (auto& p : points) {
        sf::CircleShape circle(2);
        circle.setFillColor(sf::Color::Blue);
        circle.setPosition(p.x * scaleX, size - p.y * scaleY);
        texture.draw(circle);
    }

    // Draw edges
    for (auto& e : edges) {
        sf::Vertex line[] = {
            sf::Vertex(sf::Vector2f(e.first.x * scaleX, size - e.first.y * scaleY), sf::Color::Red),
            sf::Vertex(sf::Vector2f(e.second.x * scaleX, size - e.second.y * scaleY), sf::Color::Red)
        };
        texture.draw(line, 2, sf::Lines);
    }

    texture.display();
    sf::Texture result = texture.getTexture();
    sf::Image screenshot = result.copyToImage();
    screenshot.saveToFile(filename + "_hull.png");
    std::cout << "Saved PNG: " << filename + "_hull.png\n";
}*/

// ----------------------
// Main
// ----------------------
int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage: ./hull <points_file> [visualize=1]\n";
        return 1;
    }

    std::string filename = argv[1];
    bool visualize_flag = (argc >= 3 && std::string(argv[2]) == "1");

    auto points = load_points(filename);

    if (points.empty()) {
        std::cout << "No points loaded.\n";
        return 1;
    }

    // --------- TIMING ---------
    auto start = std::chrono::high_resolution_clock::now();
    auto edges = brute_force_convex_hull(points);
    auto end = std::chrono::high_resolution_clock::now();
    double elapsed = std::chrono::duration<double>(end - start).count();
    std::cout << "Hull computation time: " << elapsed << " seconds\n";

    /*/ Optional visualization (AFTER RUNTIME ALREADY MEASURED FOR ALGORITHM)
    if (visualize_flag) {
        visualize(points, edges, filename);
    }*/

    return 0;
}

File: Code/brute_hull.py
import matplotlib.pyplot as plt
import sys
import time  # <-- add this

# ------------------------------
# Load points from text file
# ------------------------------
def load_points(filename):
    points = []
    with open(filename, "r") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            x_str, y_str = line.split(",")
            x = float(x_str.strip())
            y = float(y_str.strip())
            points.append((x, y))
    return points

# ------------------------------
# Convex Hull – Brute Force
# ------------------------------
def cross_product(p1, p2, p3):
    x1, y1 = p1
    x2, y2 = p2
    x3, y3 = p3
    return (x2 - x1)*(y3 - y1) - (y2 - y1)*(x3 - x1)

def brute_force_convex_hull(points):
    n = len(points)
    edges = []
    for i in range(n):
        for j in range(i + 1, n):
            p1 = points[i]
            p2 = points[j]
            left = right = False
            for k in range(n):
                if k == i or k == j:
                    continue
                cp = cross_product(p1, p2, points[k])
                if cp > 0:
                    left = True
                elif cp < 0:
                    right = True
                if left and right:
                    break
            if not (left and right):
                edges.append((p1, p2))
    return edges

# ------------------------------
# MAIN
# ------------------------------
if len(sys.argv) != 2:
    print("Usage: python3 brute_hull.py <filename>")
    sys.exit(1)

filename = sys.argv[1]
points = load_points(filename)

# ---- TIMING START ----
start_time = time.time()

edges = brute_force_convex_hull(points)

end_time = time.time()
elapsed = end_time - start_time
print(f"Brute-force hull completed in {elapsed:.6f} seconds")
# ---- TIMING END ----

# Visualization
plt.figure(figsize=(6,6))
x_vals, y_vals = zip(*points)
plt.scatter(x_vals, y_vals)

# Draw hull edges
for p1, p2 in edges:
    plt.plot([p1[0], p2[0]], [p1[1], p2[1]])

plt.title(f"Brute-Force Convex Hull: {filename}")
plt.xlabel("X")
plt.ylabel("Y")
plt.grid(True)

# Save figure
output_file = f"{filename}_hull.png"
plt.savefig(output_file)
print(f"Saved hull visualization as {output_file}")